### ISSUE TITLE
Warn a mistake usage of `matches!` when the "pattern" field is a variable

### ISSUE THREAD

**Issue by asuper0** [rust-lang/rust-clippy#13385]
### What it does

```rust
#[derive(Debug)]
enum TheEnum {
    AA,
    BB,
    CC,
}

pub fn test_enum() {
    let ea = TheEnum::AA;
    let m = TheEnum::BB;
    is_eq = matches!(&ea, m);
    println!("{:?} eq to {:?}: {}", ea, m, is_eq);
}
```

The result is：
```
AA eq to BB: true
```

The code above is wrong, `matches!(&ea, m)` will always return true, because "m" at the line `is_eq = matches!(&ea, m);` is not the variable `m`. I suggest to warn code like this.

### Example

```rust
matches!(&ea, m)
```

Could be written as:

```rust
ea == m
```
and derive `TheEnum` with `PartialEq`
```rust
#[derive(Debug, PartialEq)]
enum TheEnum {
```

**Comment by y21** [comment]
This has come up multiple times in the past and an implementation PR for this was also submitted, but it was closed in favor of rust-lang/rust#86112

That is, this mistake is already caught by the `unused_variables` lint, but it could be improved to explicitly point out this mistake

**Comment by asuper0** [comment]
OK, thanks. I just searched issue with title "matches" in this repository and found nothing...
